### ISSUE TITLE
Heterogeneous and missing data

### ISSUE THREAD

**Issue by NightmareNyx** [wohlert/semi-supervised-pytorch#20]
Hello,

Does any of the models work well with heterogeneous data (multiple likelihoods) and/or with missing data from the features? Are there any simple additions that could be made or models that tackle such tasks and will be added in the future?
In my mind I have the HI-VAE model more or less, but as far as I know it doesn't have a semi-supervised extension.

**Comment by wohlert** [comment]
I don't know anything about heterogeneous data or imputation, but judging from the HI-VAE paper it should be possible.

For the generative model assume that z, and label (l) are conditionally independent given x under transformations g, f.

<img src="https://latex.codecogs.com/gif.latex?p(x_n,z_n,s_n,\ell_n)=p(s_n)p(z_n|s_n)p(\ell_n|s_n)\prod_dp(x_{nd}|\gamma_{nd}),\gamma_{nd}=h_{d}(g(z_n),f(\ell_n),s_n)" />

For the recognition model we use the same assumption to infer z and l independently from each other.

<img src="https://latex.codecogs.com/gif.latex?q(s_n,z_n,\ell_n,x_m|x_n^o)=q(s_n|x_n^o)q(z_n|x_n^o,s_n)q(\ell_n|x_n^o,s_n)\prod_dp(x_{nd}|z_n,\ell_n,s_n)" title="q(s_n,z_n,\ell_n,x_m|x_n^o)=q(s_n|x_n^o)q(z_n|x_n^o,s_n)q(\ell_n|x_n^o,s_n)\prod_dp(x_{nd}|z_n,\ell_n,s_n)" />

Recognition for label will be a categorical distribution given the data x sent through a "classification network".

<img src="https://latex.codecogs.com/gif.latex?q(\ell_n|x_n^o,s_n)=\text{Cat}(\ell|\pi(x_n))" title="q(\ell_n|x_n^o,s_n)=\text{Cat}(\ell|\pi(x_n))" />

For the ELBO, you will have to figure this one out by yourself.

**Comment by xiayzh** [comment]
Hello, I have a question for the code in ladder decoder layer.
In each ladder decoder, why there are  two different NNs( self.linear1 and self.linear2) to compute mu sigma for merging and sampling,

`self.linear1 = nn.Linear(x_dim, h_dim)
 self.batchnorm1 = nn.BatchNorm1d(h_dim)
 self.merge = GaussianMerge(h_dim, self.z_dim)

self.linear2 = nn.Linear(x_dim, h_dim)
self.batchnorm2 = nn.BatchNorm1d(h_dim)
self.sample = GaussianSample(h_dim, self.z_dim)
`
`z = self.linear1(x)
z = F.leaky_relu(self.batchnorm1(z), 0.1)
q_z, q_mu, q_log_var = self.merge(z, l_mu, l_log_var)

# Sample from the decoder and send forward
 z = self.linear2(x)
z = F.leaky_relu(self.batchnorm2(z), 0.1)
z, p_mu, p_log_var = self.sample(z)`
 rather than  using tha same module
`z = self.linear1(x)
z = F.leaky_relu(self.batchnorm1(z), 0.1)
q_z, q_mu, q_log_var = self.merge(z, l_mu, l_log_var)
z, p_mu, p_log_var = self.sample(z)`

**Comment by NightmareNyx** [comment]
Thanks for the reply and the suggestion! I had something like this in mind, but wasn't sure how to properly formulate the addition of the label latent variable. Now it's clear to me. Thank you very much.
ELBO is tricky, but I'll figure it out.
